events_test.go

```
@@ -0,0 +1,424 @@
+	evsw.AddListenerForEvent("listener1", "event1",
+		func(data EventData) {
+			numbers1 <- data.(uint64)
+		})
+	evsw.AddListenerForEvent("listener1", "event2",
+		func(data EventData) {
+			numbers1 <- data.(uint64)
+		})
+	evsw.AddListenerForEvent("listener1", "event3",
+		func(data EventData) {
+			numbers1 <- data.(uint64)
+		})
+	evsw.AddListenerForEvent("listener2", "event1",
+		func(data EventData) {
+			numbers2 <- data.(uint64)
+		})
+	evsw.AddListenerForEvent("listener2", "event2",
+		func(data EventData) {
+			numbers2 <- data.(uint64)
+		})
+	evsw.AddListenerForEvent("listener2", "event3",
+		func(data EventData) {
+			numbers2 <- data.(uint64)
+		})
+	// collect received events for event1
+	go func() {
+		var sum uint64 = 0
+		for {
+			j, more := <-numbers1
+			sum += j
+			if !more {
+				doneSum1 <- sum
+				close(doneSum1)
+				return
+			}
+		}
+	}()
+	// collect received events for event2
+	go func() {
+		var sum uint64 = 0
+		for {
+			j, more := <-numbers2
+			sum += j
+			if !more {
+				doneSum2 <- sum
+				close(doneSum2)
+				return
+			}
+		}
+	}()
+	// go fire events
+	sendEvents := func(event string, doneChan chan uint64, offset uint64) {
+		var sentSum uint64 = 0
+		for i := offset; i <= offset+uint64(999); i++ {
+			sentSum += i
+			evsw.FireEvent(event, i)
+		}
+		doneChan <- sentSum
+		close(doneChan)
+		return
+	}
+	addListenersStress := func() {
+		s1 := rand.NewSource(time.Now().UnixNano())
+		r1 := rand.New(s1)
+		for k := uint16(0); k < 400; k++ {
+			listenerNumber := r1.Intn(100) + 3
+			eventNumber := r1.Intn(3) + 1
+			go evsw.AddListenerForEvent(fmt.Sprintf("listener%v", listenerNumber),
+				fmt.Sprintf("event%v", eventNumber),
+				func(_ EventData) {})
+		}
+	}
+	removeListenersStress := func() {
+		s2 := rand.NewSource(time.Now().UnixNano())
+		r2 := rand.New(s2)
+		for k := uint16(0); k < 80; k++ {
+			listenerNumber := r2.Intn(100) + 3
+			go evsw.RemoveListener(fmt.Sprintf("listener%v", listenerNumber))
```

Smart!  I would have loved to see that panic explicitly too, and maybe this was what I was missing.  Will try!